### PR TITLE
VB-3054 Update non-association event endpoint and data processing, ad…

### PR DESCRIPTION
**Pull request by AledWEvans**
VB-3054 Update non-association event endpoint and data processing
Updates NA to new events

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt [hunk 1]
```diff
@@ -20,4 +20,21 @@ interface VisitNotificationEventRepository : JpaRepository<VisitNotificationEven
     bookingReference: String,
     notificationEvent: NotificationEventType,
   ): Boolean
+
+  @Query(
+    "SELECT vne.* FROM visit_notification_event vne " +
+      " JOIN visit v on v.reference  = vne.booking_reference  " +
+      " JOIN prison p on p.id  = v.prison_id " +
+      " WHERE  v.visit_start >= NOW() " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt:28]: a suggestion, can be ignored - we could pass the date as a parameter.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt [hunk 2]
```diff
@@ -33,9 +37,16 @@ class VisitNotificationEventService(
...
   fun handleNonAssociations(notificationDto: NonAssociationChangedNotificationDto) {
-    if (prisonerService.hasPrisonerGotANonAssociationWith(notificationDto.prisonerNumber, notificationDto.nonAssociationPrisonerNumber)) {
-      val affectedVisits = getOverLappingVisits(notificationDto)
-      processVisitsWithNotifications(affectedVisits, NON_ASSOCIATION_EVENT)
+    with(notificationDto) {
+      if (NON_ASSOCIATION_CREATED == type) {
+        val affectedVisits = getOverLappingVisits(notificationDto)
+        processVisitsWithNotifications(affectedVisits, NON_ASSOCIATION_EVENT)
+      } else if (type in arrayOf(NON_ASSOCIATION_DELETED, NON_ASSOCIATION_CLOSED)) {
+        if (!prisonerService.hasPrisonerGotANonAssociationWith(prisonerNumber, nonAssociationPrisonerNumber)) {
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:45]: is the ! (not) incorrect?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt [hunk 3]
```diff
@@ -104,16 +119,40 @@ class VisitNotificationEventService(
...
+    val prisonerEventsToBeDeleted = prisonersNotifications.filter {
+        prisonersNotify ->
+      nonAssociationPrisonersNotifications.any { it.reference == prisonersNotify.reference }
+    }
+    val nsPrisonerEventsToBeDeleted = nonAssociationPrisonersNotifications.filter {
+        nsPrisonersNotify ->
+      prisonerEventsToBeDeleted.any { it.reference == nsPrisonersNotify.reference }
+    }
+
+    return (prisonerEventsToBeDeleted + nsPrisonerEventsToBeDeleted).sortedBy { it.reference }
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:148]: is a sort needed here? Does not matter though if it stays.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt [hunk 4]
```diff
@@ -104,16 +119,40 @@ class VisitNotificationEventService(
...
+  }
+
+  private fun getPrisonCode(prisonerNumber: String): String {
+    return prisonerService.getPrisonerPrisonCode(prisonerNumber)
+      ?: throw ItemNotFoundException("Prison can not be found for prisoner: $prisonerNumber")
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:153]: If we get a notification for a non supported prisoner will ItemNotFoundException be thrown? Will the message end up in DLQ as a result?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt:28]: I thought about this, there is no requirement for this, you always pick up on this :-)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:45]: Yes, basically if there is no non associations it means we can safely delete, other wise there could muliple notifications for mulitile non associations for the same visits and we don't know which one!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:148]: well it's just encase,.. probably okay.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt (hunk 4 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:153]: depends how we handling it orch service, not the responsibility of this code, but good point not will look into the orch side of things.

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:153]: may be return another code?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt [hunk 5]
```diff
@@ -119,19 +122,16 @@ class VisitNotificationEventService(
...
     val fromDate = LocalDate.now().atStartOfDay()
 
     val primaryPrisonerVisits = visitService.getBookedVisits(notificationDto.prisonerNumber, prisonCode, fromDate)
     val nonAssociationPrisonerVisits = visitService.getBookedVisits(notificationDto.nonAssociationPrisonerNumber, prisonCode, fromDate)
     return getOverLappingVisits(primaryPrisonerVisits, nonAssociationPrisonerVisits)
   }
 
-  private fun getAffectedNotifications(notificationDto: NonAssociationChangedNotificationDto): List<VisitNotificationEvent> {
+  private fun getAffectedNotifications(notificationDto: NonAssociationChangedNotificationDto, prisonCode: String): List<VisitNotificationEvent> {
     // get the prisoners' prison code
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationEventService.kt:134]: minor - this can go

**Review by dhirajshettymoj — APPROVED** [review]

**Pull request merged**